python/_tskitmodule.c

```diff
@@ -11295,6 +11301,221 @@ static PyTypeObject VariantGeneratorType = {
+        goto out;
+    }
+    if (samples_input != Py_None) {
+        samples_array = (PyArrayObject *) PyArray_FROMANY(
+            samples_input, NPY_INT32, 1, 1, NPY_ARRAY_IN_ARRAY);
+        if (samples_array == NULL) {
+            goto out;
+        }
+        shape = PyArray_DIMS(samples_array);
+        num_samples = (tsk_size_t) shape[0];
+        samples = PyArray_DATA(samples_array);
+    }
+    if (py_alleles != Py_None) {
+        alleles = parse_allele_list(py_alleles);
+        if (alleles == NULL) {
+            goto out;
+        }
+    }
+    self->variant = PyMem_Malloc(sizeof(tsk_vargen_t));
+    if (self->variant == NULL) {
+        PyErr_NoMemory();
+        goto out;
+    }
+    /* Note: the variant currently takes a copy of the samples list. If we wanted
+     * to avoid this we would INCREF the samples array above and keep a reference
+     * to in the object struct */
+    err = tsk_variant_init(self->variant, self->tree_sequence->tree_sequence, samples,
+        num_samples, alleles, options);
+    if (err != 0) {
+        handle_library_error(err);
+        goto out;
+    }
+    ret = 0;
+out:
+    PyMem_Free(alleles);
+    Py_XDECREF(samples_array);
+    return ret;
+}
+
+static PyObject *
+Variant_decode(Variant *self, PyObject *args, PyObject *kwds)
+{
+    int err;
+    PyObject *ret = NULL;
+    tsk_id_t site_id;
+    static char *kwlist[] = { "site", NULL };
+
+    if (Variant_check_state(self) != 0) {
+        goto out;
+    }
+    if (!PyArg_ParseTupleAndKeywords(
+            args, kwds, "O&", kwlist, &tsk_id_converter, &site_id)) {
+        goto out;
+    }
+    err = tsk_variant_decode(self->variant, site_id, 0);
+    if (err != 0) {
+        handle_library_error(err);
+        goto out;
+    }
+
+    ret = Py_BuildValue("");
+out:
+    return ret;
+}
+
+static PyObject *
+Variant_get_site_id(Variant *self, void *closure)
+{
+    PyObject *ret = NULL;
+    if (Variant_check_state(self) != 0) {
+        goto out;
+    }
+    ret = Py_BuildValue("n", (Py_ssize_t) self->variant->site.id);
+out:
+    return ret;
+}
+
+static PyObject *
+Variant_get_alleles(Variant *self, void *closure)
+{
+    PyObject *ret = NULL;
+
+    if (Variant_check_state(self) != 0) {
+        goto out;
+    }
+    ret = make_alleles(self->variant);
+out:
+    return ret;
+}
+
+static PyObject *
+Variant_get_genotypes(Variant *self, void *closure)
+{
+    PyObject *ret = NULL;
+    PyArrayObject *array = NULL;
+    npy_intp dims;
+
+    if (Variant_check_state(self) != 0) {
+        goto out;
+    }
+
+    dims = self->variant->num_samples;
+    array = (PyArrayObject *) PyArray_SimpleNewFromData(
+        1, &dims, NPY_INT32, self->variant->genotypes);
+    if (array == NULL) {
+        goto out;
+    }
+    PyArray_CLEARFLAGS(array, NPY_ARRAY_WRITEABLE);
+    if (PyArray_SetBaseObject(array, (PyObject *) self) != 0) {
+        goto out;
+    }
+    /* PyArray_SetBaseObject steals a reference, so we have to incref the tree
```

Comment:
out of date comment here, no tree

Reply:
Now you know where I copy pasted from :smiley:! Fixed,